Increased documentation content of README and root rustdoc page. (#296)
georust/gdal: Readme.md

# GDAL

[![Documentation](https://docs.rs/gdal/badge.svg)](https://docs.rs/gdal)
![Build Status](https://github.com/georust/gdal/workflows/CI/badge.svg)

[GDAL](http://gdal.org/) is a translator and processing library for various raster and vector geospatial data formats. 

This crate provides safe, idiomatic [Rust](http://www.rust-lang.org/) bindings for GDAL.

## Capabilities

GDAL is an incredibly powerful library. For a general understanding of its capabilities, a good place to get started is the [GDAL User-oriented documentation](https://gdal.org/user/index.html). These features include:

> * Open a raster dataset for reading/writing

Reviewer: Nit: I think the continuous form sounds better here and below:

```suggestion
* Opening a raster dataset for reading/writing
```

* Get the raster size and number of bands
* Get and set the projection and geo-transform
* Warp (resample and reproject) between datasets
* Read and write vector data
* Read and write metadata
* Translate between formats

## Documentation

> This crate's [API documentation](https://docs.rs/crate/gdal) is hosted on [docs.rs](https://docs.rs). The Rust documentation is currently a work in progress, and may not cover requisite details on parameter semantics, value interpretation, etc.. Therefore, the authoritative documentation for the [GDAL C API](https://gdal.org/api/index.html#c-api) may be a helpful reference.

Reviewer: It might be worth linking to the C++ API too, since it's matches the Rust one better and usually has more details.

Author: @lnicola I made this point about C vs C++ because when I was new to the crate, I found it confusing on where to go for semantic details, especially when some of the names are the same or similar. I assumed we only link to C because that's what `bindgen` provides for us, and then there's larger complexities with name-mangling, memory model, libraries, etc. in C++. 

Were you thinking I should strike this section, or refine it, or was this just a general point?

Reviewer: Of course, we're calling the C API, but ours is more similar in spirit with the C++ one. Anyway, my point was about the documentation: the C GDAL docs are pretty bare, while the C++ ones have more details.

Compare https://gdal.org/doxygen/gdal_8h.html#a764cb93a266ada902b13c2bdad3f7037 with https://gdal.org/doxygen/classGDALDataset.html#ad2a643880bf8b3c492783e1ab5fbc667.

So I was thinking we could put it more like:

```
Therefore, the authoritative documentation is that of GDAL in the form of its [C](https://gdal.org/api/index.html#c-api) and [C++](https://gdal.org/api/index.html#id3) APIs.
The latter is usually better documented.
```

Author: I think I misread your comment... more clear now that I came back to it. My point is somewhat relevant still; when I was just coming to this crate, I would try to translate from the C++ docs to Rust, and hit a wall because we don't implement the C++ functionality, and didn't to save folks from that pothole.

Reviewer: > because we don't implement the C++ functionality

Sure, but our API doesn't look like the C one either :smile:.

## Usage

> This crate provides idiomatic Rust bindings for the GDAL C library. To do that, it internally uses [`gdal-sys`](gdal-sys), which is a  [`bindgen`](https://rust-lang.github.io/rust-bindgen/)-generated sub-crate, rendered directly from the GDAL C library. Normally one would not need to interact with the `gdal-sys` crate. However, not every single feature of the GDAL C API has been fully exposed, so in some cases the `gdal-sys` crate may be necessary to accessing more esoteric features of GDAL.

Reviewer: ```suggestion
This crate provides high-level, idiomatic Rust bindings for GDAL.
To do that, it internally uses [`gdal-sys`](gdal-sys), a  low-level interface to the GDAL C library, generated using [`bindgen`](https://rust-lang.github.io/rust-bindgen/).
Using the `gdal-sys` crate is normally not needed, but it can be useful in order to call APIs that have not yet been exposed in `gdal`.
```

> _Note_: GDAL offers both C and C++ APIs. This crate only makes use of the GDAL C API.

> Building this crate assumes a compatible version of GDAL is installed with requisite header files and shared libraries. This repository includes pre-generated bindings for GDAL 2.4 through 3.5 (see the`gdal-sys/prebuilt-bindings` directory). If you're compiling against a later version of GDAL, you can enable the `bindgen` feature flag to have new bindings generated on the fly.  

Reviewer: ```suggestion
Building this crate assumes a compatible version of GDAL is installed with the corresponding header files and shared libraries.
This repository includes pre-generated bindings for GDAL 2.4 through 3.5 (see the`gdal-sys/prebuilt-bindings` directory).
If you're compiling against a later version of GDAL, you can enable the `bindgen` feature flag to have new bindings generated on the fly.
```

## Community

This crate is part of the expansive (and expanding!) [`georust`](https://georust.org/) organization. Come join our discussions on [Discord](https://discord.gg/Fp2aape)!

## Contributing

> This crate continues to evolve, and PRs are welcome. Make sure you are comfortable with the [Code of Conduct](CODE_OF_CONDUCT.md) and [License](LICENSE.txt) and before submitting a PR.

Reviewer 2: ```suggestion
This crate continues to evolve, and PRs are welcome. Make sure you are comfortable with the [Code of Conduct](CODE_OF_CONDUCT.md) and [License](LICENSE.txt) before submitting a PR.
```
